compiler/src/iree/compiler/API/Internal/Embed.cpp

```diff
@@ -603,14 +644,63 @@ const char *ireeCompilerErrorGetMessage(iree_compiler_error_t *error) {
 
 int ireeCompilerGetAPIVersion() { return 0; }
 
-void ireeCompilerGlobalInitialize(bool initializeCommandLine) {
-  if (globalInit || isShutdown) {
-    fprintf(
-        stderr,
-        "FATAL ERROR: ireeCompilerGlobalInitialize called multiple times\n");
+void ireeCompilerGetProcessCLArgs(int *argc, const char ***argv) {
+#ifdef _WIN32
+  // See the Windows command line processing in InitLLVM.cpp. It hasn't
+  // changed in forever.
+  std::string banner = std::string(argv[0]) + ": ";
+  ExitOnError(banner);
+  ExitOnError(errorCodeToError(
+      windows::GetCommandLineArguments(globalInit->retainedArgv, alloc)));
+  // GetCommandLineArguments doesn't terminate with a nullptr per what argv
+  // expects, so do that.
+  globalInit->retainedArgv.push_back(nullptr);
+  *argc = globalInit->retainedArgv.size() - 1;
+  *argv = globalInit->retainedArgv.data();
+#endif
+}
+
+void ireeCompilerSetupGlobalCL(int argc, const char **argv, const char *banner,
+                               bool installSignalHandlers) {
+  if (globalInit->usesCommandLine) {
+    fprintf(stderr, "FATAL ERROR: ireeCompileParseCL called multiple times\n");
+    abort();
+  }
+  globalInit->usesCommandLine = true;
+  globalInit->registerCommandLineOptions();
+
+  llvm::setBugReportMsg(
+      "Please report issues to https://github.com/openxla/iree/issues and "
+      "include the crash backtrace.\n");
+  llvm::cl::SetVersionPrinter(llvmVersionPrinter);
+
+  if (installSignalHandlers) {
+    // A few other things from InitLLVM to setup default command-line signal
+    // handlers. See InitLLVM::InitLLVM for the initialization sequence and
+    // commentary, should it ever be necessary to revist this (it hasn't changed
+    // in many years).
+    llvm::sys::SetOneShotPipeSignalFunction(
+        llvm::sys::DefaultOneShotPipeSignalHandler);
+    static llvm::PrettyStackTraceProgram stackPrinter(argc, argv);
+    llvm::sys::PrintStackTraceOnErrorSignal(argv[0]);
+    llvm::install_out_of_memory_new_handler();
+  }
+
+  llvm::cl::ParseCommandLineOptions(argc, argv, banner);
+}
+
+void ireeCompilerGlobalInitialize() {
+  if (globalInit) {
+    globalInit->refCount.fetch_add(1);
+    return;
+  }
+  if (isShutdown) {
+    fprintf(stderr,
+            "FATAL ERROR: ireeCompilerGlobalInitialize called multiple times "
```

Comment:
It can be more specific saying that ireeCompilerGlobalInitialize called while shutting down another.